### ISSUE TITLE
[SC2317] False positive with nested functions

### ISSUE THREAD

**Issue by GrabbenD** [koalaman/shellcheck#3057]
#### For bugs
- Rule Id = [SC2317](https://www.shellcheck.net/wiki/SC2317)
- My shellcheck version = `0.10.0`
- [x] The rule's wiki page does not already cover this
- [x] I tried on https://www.shellcheck.net/

---

#### Non-problematic code example

```sh
#!/usr/bin/env bash

PARENT () {
  echo 'test'

  NESTED () {
    echo 'ok'
  }
}

PARENT
```

---

#### Here's a snippet or screenshot that shows the problem:

When function is exported:

```sh
PARENT () {
  echo 'test'

  NESTED () {
    echo 'SC2317 ERROR'
  }
}

func=$(declare -f PARENT)
$func  # PARENT echo works
NESTED # NESTED echo works
```

When exported function is called in a new shell

```sh
PARENT () {
  echo 'test'

  NESTED () {
    echo 'SC2317 ERROR'
  }
}

cmd="$(declare -f PARENT); PARENT; NESTED"
runuser --login "${USER}" --shell "/bin/bash" -- -c "${cmd}" # Both echo works
#bash -c "${cmd}"
```

When function is called indirectly as a function argument:

```sh
PARENT () {
  echo 'test'

  NESTED () {
    echo 'SC2317 ERROR'
  }
}

HANDLER () {
  $1     # PARENT echo works
  NESTED # NESTED echo works
}

HANDLER PARENT
```

---

#### Here's what shellcheck currently says:

`www.shellcheck.net`
> [SC2329](https://www.shellcheck.net/wiki/SC2329) (info): This function is never invoked. Check usage (or ignored if invoked indirectly).

`VSCode`
> Command appears to be unreachable. Check usage (or ignore if invoked indirectly). shellcheck([SC2329](https://www.shellcheck.net/wiki/SC2329))

---

#### Here's what I wanted or expected to see:

Indirectly invoked nested functions shouldn't trigger `SC2329`
